nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -572,197 +564,50 @@ impl DataStore {
-                        .await
-                        .optional()?;
-
-                    let volume = if let Some(v) = volume {
-                        v
-                    } else {
-                        // the volume was hard-deleted, return an empty
-                        // CrucibleResources
-                        return Ok(CrucibleResources::V1(
-                            CrucibleResourcesV1::default(),
-                        ));
-                    };
+        let _old_volume: Vec<Volume> =
+            DecreaseCrucibleResourceCountAndSoftDeleteVolume::new(
+                volume_id,
+                crucible_targets.read_only_targets.clone(),
+            )
+            .get_results_async::<Volume>(
+                &*self.pool_connection_unauthorized().await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
 
-                    if volume.time_deleted.is_none() {
-                        // a volume record exists, and was not deleted - this is the
-                        // first time through this transaction for a certain volume
-                        // id. Get the volume for use in the transaction.
-                        volume
-                    } else {
-                        // this volume was soft deleted - this is a repeat time
-                        // through this transaction.
-
-                        if let Some(resources_to_clean_up) =
-                            volume.resources_to_clean_up
-                        {
-                            // return the serialized CrucibleResources
-                            return serde_json::from_str(
-                                &resources_to_clean_up,
-                            )
-                            .map_err(|e| {
-                                TxnError::CustomError(
-                                    DecreaseCrucibleResourcesError::SerdeError(
-                                        e,
-                                    ),
-                                )
-                            });
-                        } else {
-                            // If no CrucibleResources struct was serialized, that's
-                            // definitely a bug of some sort - the soft-delete below
-                            // sets time_deleted at the same time as
-                            // resources_to_clean_up! But, instead of a panic here,
-                            // just return an empty CrucibleResources.
-                            return Ok(CrucibleResources::V1(
-                                CrucibleResourcesV1::default(),
-                            ));
-                        }
+        // Get the updated Volume to get the resources to clean up.
+        // TODO figure out how DecreaseCrucibleResourceCountAndSoftDeleteVolume
+        // can return the updated volume in away that Diesel likes
```

Comment:
Is Diesel the problem, or the CTE itself?

When writing CTEs that do modification and want to return the "new" result, I usually need to synthetically construct the output - but that's kinda a SQL problem, not really a Diesel problem, because "SELECT"-ing the old value returns what it was before the CTE ran.

Either way, happy to help figure out how to make this happen.

Reply:
That was written when the last SELECT in the CTE looked like the following (to return the updated volume instead): 

```
select case
 when volume.resources_to_clean_up is not null then volume.*
 else (select * from <CTE temporary table with the same schema as volume> where id = '<volume id>')
end
from volume where id = '<volume id>';
```
where (I assume) Diesel was having trouble with the `select * from <CTE TEMPORARY TABLE>` -> `Vec<Volume>`:

```
03:30:28.894Z DEBG 9c4e4732-8af8-4267-9da7-8331a0e8aa86 (ServerContext): recording saga event
    event_type = Failed(ActionFailed { source_error: String("failed to decrease_crucible_resource_count_and_soft_delete_volume: InternalError { internal_message: \\"unexpected database error: invalid length: expected 16 bytes, found 1118\\" }") })
    node_id = 5
    saga_id = ca596881-4f83-4288-a79b-dade5377206c
```

Manually calling this CTE in CRDB shows that one column is being returned as a large tuple in one column titled "case", which is probably the source of the invalid length error.

The CTE in this PR was changed to end with selecting `volume.*` where the id matches the argument volume id, returning the old volume.

Reply:
Here's the diff to recreate:

```patch
diff --git a/nexus/db-queries/src/db/queries/volume.rs b/nexus/db-queries/src/db/queries/volume.rs
index 31882dca8..a34b78d18 100644
--- a/nexus/db-queries/src/db/queries/volume.rs
+++ b/nexus/db-queries/src/db/queries/volume.rs
@@ -460,11 +460,27 @@ impl QueryFragment<Pg> for DecreaseCrucibleResourceCountAndSoftDeleteVolume {
         self.conditionally_update_volume_query.walk_ast(out.reborrow())?;
         out.push_sql(") ");
 
-        out.push_sql("SELECT ");
+        //out.push_sql("SELECT ");
+        //dsl::volume.walk_ast(out.reborrow())?;
+        //out.push_sql(".* FROM ");
+        //dsl::volume.walk_ast(out.reborrow())?;
+        //out.push_sql(" WHERE ");
+        //out.push_identifier(dsl::id::NAME)?;
+        //out.push_sql(" = ");
+        //out.push_bind_param::<sql_types::Uuid, Uuid>(&self.volume_id)?;
+
+        out.push_sql("SELECT CASE WHEN ");
+        out.push_identifier(dsl::resources_to_clean_up::NAME)?;
+        out.push_sql(" is not null then ");
         dsl::volume.walk_ast(out.reborrow())?;
-        out.push_sql(".* FROM ");
+        out.push_sql(".* else (");
+        out.push_sql("select * from ");
+        out.push_sql(Self::UPDATED_VOLUME_TABLE);
+        out.push_sql(" where id = ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.volume_id)?;
+        out.push_sql(") end from ");
         dsl::volume.walk_ast(out.reborrow())?;
-        out.push_sql(" WHERE ");
+        out.push_sql(" where ");
         out.push_identifier(dsl::id::NAME)?;
         out.push_sql(" = ");
         out.push_bind_param::<sql_types::Uuid, Uuid>(&self.volume_id)?;
```

Reply:
I've removed that comment in 30d3ed109